File: VB/SLGridExample/GridDetailContainer.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Net
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Documents
Imports System.Windows.Ink
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Animation
Imports System.Windows.Shapes
Imports DevExpress.Xpf.Grid
Imports System.Windows.Data

Namespace SLGridExample
	Public Class GridDetailContainer
		Inherits ContentPresenter

		Public Shared ReadOnly IsVisibleProperty As DependencyProperty = DependencyProperty.Register("IsVisible", GetType(Boolean), GetType(GridDetailContainer), New PropertyMetadata(False, New PropertyChangedCallback(AddressOf GridPropertyChangedCallback)))

		Public Overloads Property IsVisible() As Boolean
			Get
				Return CBool(GetValue(IsVisibleProperty))
			End Get
			Set(ByVal value As Boolean)
				SetValue(IsVisibleProperty, value)
			End Set
		End Property

		Public Shared ReadOnly GridProperty As DependencyProperty = DependencyProperty.Register("Grid", GetType(GridControl), GetType(GridDetailContainer), New PropertyMetadata(Nothing, New PropertyChangedCallback(AddressOf GridPropertyChangedCallback)))

		Public Property Grid() As GridControl
			Get
				Return CType(GetValue(GridProperty), GridControl)
			End Get
			Set(ByVal value As GridControl)
				SetValue(GridProperty, value)
			End Set
		End Property

		Private Shared Sub GridPropertyChangedCallback(ByVal s As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			CType(s, GridDetailContainer).OnChanged()
		End Sub

		Private gridItemsSourceBinding_Renamed As Binding
		Public Property GridItemsSourceBinding() As Binding
			Get
				Return gridItemsSourceBinding_Renamed
			End Get
			Set(ByVal value As Binding)
				gridItemsSourceBinding_Renamed = value
				OnChanged()
			End Set
		End Property

		Private Sub OnChanged()
			Content = Grid
			If IsVisible Then
				Visibility = Visibility.Visible
				If Grid IsNot Nothing AndAlso GridItemsSourceBinding IsNot Nothing Then
					Grid.SetBinding(GridControl.DataSourceProperty, GridItemsSourceBinding)
				End If
			Else
				Visibility = Visibility.Collapsed
				If Grid IsNot Nothing Then
					Grid.ClearValue(GridControl.DataSourceProperty)
				End If
			End If
		End Sub

	End Class
End Namespace

File: CS/SLGridExample/GridDetailContainer.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using DevExpress.Xpf.Grid;
using System.Windows.Data;

namespace SLGridExample {
    public class GridDetailContainer : ContentPresenter {

        public static readonly DependencyProperty IsVisibleProperty =
            DependencyProperty.Register("IsVisible", typeof(bool), typeof(GridDetailContainer), new PropertyMetadata(false, new PropertyChangedCallback(GridPropertyChangedCallback)));

        public bool IsVisible {
            get { return (bool)GetValue(IsVisibleProperty); }
            set { SetValue(IsVisibleProperty, value); }
        }

        public static readonly DependencyProperty GridProperty =
            DependencyProperty.Register("Grid", typeof(GridControl), typeof(GridDetailContainer), new PropertyMetadata(null, new PropertyChangedCallback(GridPropertyChangedCallback)));

        public GridControl Grid {
            get { return (GridControl)GetValue(GridProperty); }
            set { SetValue(GridProperty, value); }
        }

        static void GridPropertyChangedCallback(DependencyObject s, DependencyPropertyChangedEventArgs e) {
            ((GridDetailContainer)s).OnChanged();
        }

        Binding gridItemsSourceBinding;
        public Binding GridItemsSourceBinding {
            get { return gridItemsSourceBinding; }
            set {
                gridItemsSourceBinding = value;
                OnChanged();
            }
        }

        void OnChanged() {
            Content = Grid;
            if (IsVisible) {
                Visibility = Visibility.Visible;
                if (Grid != null && GridItemsSourceBinding != null)
                    Grid.SetBinding(GridControl.DataSourceProperty, GridItemsSourceBinding);
            } else {
                Visibility = Visibility.Collapsed;
                if (Grid != null)
                    Grid.ClearValue(GridControl.DataSourceProperty);
            }
        }

    }
}
